libraries/core_libs/consensus/src/final_chain/final_chain.cpp

```diff
@@ -32,62 +30,101 @@ class FinalChainImpl final : public FinalChain {
     return ret;
   }();
   EthBlockNumber delegation_delay_;
+
+  ValueByBlockCache<std::shared_ptr<const BlockHeader>> block_headers_cache_;
+  ValueByBlockCache<std::optional<const h256>> block_hashes_cache_;
+  ValueByBlockCache<const SharedTransactions> transactions_cache_;
+  ValueByBlockCache<std::shared_ptr<const TransactionHashes>> transaction_hashes_cache_;
+  MapByBlockCache<addr_t, std::optional<const state_api::Account>> accounts_cache_;
+
+  ValueByBlockCache<uint64_t> total_vote_count_cache_;
+  MapByBlockCache<addr_t, uint64_t> dpos_vote_count_cache_;
+  MapByBlockCache<addr_t, uint64_t> dpos_is_eligible_cache_;
+
   LOG_OBJECTS_DEFINE
 
  public:
-  FinalChainImpl(const std::shared_ptr<DB>& db, const taraxa::ChainConfig& config, const addr_t& node_addr)
+  FinalChainImpl(const std::shared_ptr<DB>& db, const taraxa::FullNodeConfig& config, const addr_t& node_addr)
       : db_(db),
-        commitee_size_(config.pbft.committee_size),
+        commitee_size_(config.chain.pbft.committee_size),
         // replay_protection_service_(NewReplayProtectionService({}, db)),
         state_api_([this](auto n) { return block_hash(n).value_or(ZeroHash()); },  //
-                   config.final_chain.state,
+                   config.chain.final_chain.state,
                    {
                        1500,
                        4,
                    },
                    {
                        db->stateDbStoragePath().string(),
-                   }) {
+                   }),
+        block_headers_cache_(config.final_chain_cache_in_blocks,
+                             [this](uint64_t blk) { return get_block_header(blk); }),
+        block_hashes_cache_(config.final_chain_cache_in_blocks, [this](uint64_t blk) { return get_block_hash(blk); }),
+        transactions_cache_(config.final_chain_cache_in_blocks, [this](uint64_t blk) { return get_transactions(blk); }),
+        transaction_hashes_cache_(config.final_chain_cache_in_blocks,
+                                  [this](uint64_t blk) { return get_transaction_hashes(blk); }),
+        accounts_cache_(config.final_chain_cache_in_blocks,
+                        [this](uint64_t blk, const addr_t& addr) { return state_api_.get_account(blk, addr); }),
+
+        total_vote_count_cache_(config.final_chain_cache_in_blocks,
+                                [this](uint64_t blk) { return get_dpos_eligible_total_vote_count(blk); }),
+        dpos_vote_count_cache_(
+            config.final_chain_cache_in_blocks,
+            [this](uint64_t blk, const addr_t& addr) { return get_dpos_eligible_vote_count(blk, addr); }),
+        dpos_is_eligible_cache_(config.final_chain_cache_in_blocks,
+                                [this](uint64_t blk, const addr_t& addr) { return get_dpos_is_eligible(blk, addr); }) {
```

Comment:
what do we need this get_dpos_is_elogible function ? cannot we use here directly return state_api_.dpos_is_eligible(blk_num, addr) ?